src/services/classifier2020.ts

```diff
@@ -0,0 +1,252 @@
+        if (decl && isBindingElement(decl)) {
+            decl = getDeclarationForBindingElement(decl);
+        }
+        return decl && tokenFromDeclarationMapping[decl.kind];
+    }
+
+    function reclassifyByType(typeChecker: TypeChecker, node: Node, typeIdx: TokenType): TokenType {
+        // type based classifications
+        if (typeIdx === TokenType.variable || typeIdx === TokenType.property || typeIdx === TokenType.parameter) {
+            const type = typeChecker.getTypeAtLocation(node);
+            if (type) {
+                const test = (condition: (type: Type) => boolean) => {
+                    return condition(type) || type.isUnion() && type.types.some(condition);
+                };
+                if (typeIdx !== TokenType.parameter && test(t => t.getConstructSignatures().length > 0)) {
+                    return TokenType.class;
+                }
+                if (test(t => t.getCallSignatures().length > 0) && !test(t => t.getProperties().length > 0) || isExpressionInCallExpression(node)) {
+                    return typeIdx === TokenType.property ? TokenType.member : TokenType.function;
+                }
+            }
+        }
+        return typeIdx;
+    }
+
+    function isLocalDeclaration(decl: Declaration, sourceFile: SourceFile): boolean {
+        if (isBindingElement(decl)) {
+            decl = getDeclarationForBindingElement(decl);
+        }
+        if (isVariableDeclaration(decl)) {
+            return (!isSourceFile(decl.parent.parent.parent) || isCatchClause(decl.parent)) && decl.getSourceFile() === sourceFile;
+        }
+        else if (isFunctionDeclaration(decl)) {
+            return !isSourceFile(decl.parent) && decl.getSourceFile() === sourceFile;
+        }
+        return false;
+    }
+
+    function getDeclarationForBindingElement(element: BindingElement): VariableDeclaration | ParameterDeclaration {
+        while (true) {
+            if (isBindingElement(element.parent.parent)) {
+                element = element.parent.parent;
+            }
+            else {
+                return element.parent.parent;
+            }
+        }
+    }
+
+    function inImportClause(node: Node): boolean {
+        const parent = node.parent;
+        return parent && (isImportClause(parent) || isImportSpecifier(parent) || isNamespaceImport(parent));
+    }
+
+    function isExpressionInCallExpression(node: Node): boolean {
+        while (isRightSideOfQualifiedNameOrPropertyAccess(node)) {
+            node = node.parent;
+        }
+        return isCallExpression(node.parent) && node.parent.expression === node;
+    }
+
+    function isRightSideOfQualifiedNameOrPropertyAccess(node: Node): boolean {
+        return (isQualifiedName(node.parent) && node.parent.right === node) || (isPropertyAccessExpression(node.parent) && node.parent.name === node);
+    }
+
+    const tokenFromDeclarationMapping: { [name: string]: TokenType } = {

```

Comment:
Should this be a `Map` (nee. `ESMap` internally)? If it's an object and the property accesses on the use sites above are possible inline cache misses, it can result in deoptimizations.

Reply:
Thanks 0 I've used a `new Map` as I found a few of those in the codebase and I assume this is fine?